4.0.0 - Huge refactor, more examples,  authentication, configurable MIMETypes

michalpokusa opened the pull request:
New Features 🎉:
    - Added `Basic` and `Bearer` authentication options on server and per route level
    - Rewritten logic for `MIMETypes`, now it is possible to remove unused, thus saving memory, and register new ones
    - `root_path` in `Server` constructor is now optional, if not specified or `None`, access to filesystem is disabled, which resolves #52 
    - `request` parameter in handler functions now has access to `server` that received it
    - EDIT1: `Request.json()` for easy parsing of JSON POST data
    - EDIT1: `host` and `port` attributes on `Server`
    - EDIT1: Debug mode for displaying messages useful during development
    - EDIT1: Added `append_slash` to `Server.route`
    - EDIT2: Added `...` and `....` wildcards
    - EDIT3: New classes fro reponse types: `ChunkedResponse`, `JSONResponse`, `FileResponse`, `Redirect`
    - Added ability to stop and start a `Server` again
    - Documented option to host multiple servers on single board, despite not being the intended use, it is possible and might be useful in some cases

Modified Features 🏗️:
    - `server.route()` now can accept list of methods in addition to single one
    - Routes now respect trailing `/` at the end of a path, so `/example` and `/example/` are two diffrent paths
    - `serve_forever` does not raise any error and continues to work no matter what
    - ~Response.send_file default to root_path instead of root of the filesystem~
    - EDIT3: `Response` is now separated into many classes for diffrent usages like sending file or JSON data
    - EDIT3: `Response` objects are now returned from handler functions instead of being used in context managers, this makes them easier and more intuitive to use, repalced all public methods with classes that have same functionality
    - URL parameters are now __keyword__ instead of __positional__, which makes them much more flexible

Refactor ⚙️:
    - **Removed `HTTP...` from class names in order to unify naming with other libs, so e.g. `HTTPServer` becomes `Server`, `HTTPRequest` becomes `Request` etc.**
    - It is now possible to import directly from `adafruit_httpserver` instead of using multiple import statements for each component
    - Replaced `CommonHTTPStatus` with objects of `Status` class
    - EDIT1: Refactor of `_Routes` and `_Route` classes

Docs 📘:
    - Updated docs to new API
    - Added new examples that show previously unexplained usages
    - Added `:emphasize-lines:` to examples to mark most important parts of them
    
Testing of these changes would be greatly appreciated 😉

michalpokusa commented:
Just out of curiosity I tried starting multiple servers on one board at the same time, of couse on diffrent ports.
Although it was not an intended use, it seems to work fine 😂.

Not sure whether it is worth documenting...

FoamyGuy commented:
I tested this out today on a Feather S2 TFT. It seems to work great, I do have a few changes to suggest:

- If the user doesn't have proper CIRCUITPY_WIFI_SSID or CIRCUITPY_WIFI_PASSWORD and they try to run in automatic mode raise some error telling them the values aren't set.
- Add a argument or flag of some sort that allows serve_forever() to print exceptions if they occur to make troubleshooting route handlers easier
- I tried to request from `/implicit-require/` and got a not found type response I think. If I understand correctly it is supposed to allow the trailing slash.
- serve_forever() should be able (even if opionally) to re-raise the `KeyboardInterrupt` exception so that user can stop with ctrl c

michalpokusa commented:
> I tested this out today on a Feather S2 TFT. It seems to work great, I do have a few changes to suggest:
> 
> * If the user doesn't have proper CIRCUITPY_WIFI_SSID or CIRCUITPY_WIFI_PASSWORD and they try to run in automatic mode raise some error telling them the values aren't set.
> * Add a argument or flag of some sort that allows serve_forever() to print exceptions if they occur to make troubleshooting route handlers easier
> * I tried to request from `/implicit-require/` and got a not found type response I think. If I understand correctly it is supposed to allow the trailing slash.
> * serve_forever() should be able (even if opionally) to re-raise the `KeyboardInterrupt` exception so that user can stop with ctrl c

Thank you for testing. According to what you pointed out I made the following changes:
- I implemented simple validation using `.getaddinfo()` that checks whether it is possible to start the server on given `host:port`.
- Added Debug mode for displaying exceptions during `.serve_forever()`, on `.start()` and `.stop()` and on incomming request
- Fixed the `KeyboardInterrupt` during `.serve_forever`

When it comes to `/implicit-require/` I saw during your stream the way you were trying to access that, because handler was only routing `/implicit-require`, the server didn't find a handler, then tried using it as a directory name and that is why it showed "No file...". The change I made allows setting diffrent routes for e.g. `/implicit-require/` and `/implicit-require` as technically they are not the same URL. Although many frameworks (like Django that you were mentioning during stream) redirect it, it is not the default behaviour. I will add `append_slash` option to allow this redirecting.

michalpokusa commented:
During a recent Deep Dive stream from @FoamyGuy, @Neradoc suggested implementing default handlers for specific error types.

Current implementation does not allow that, I am now working on rewrite of `response.py` that would change the behaviour of lib, so that a `Response` is returned from handler, which is definitely a more intuitive approach that current context manager version. This change would allow implementing default handlers and it would be **backwards compatible**, e.g. as version 4.1.0.

This is a moderately large change and it requires a lot of changes in docs, examples and nearly a complete rewrite of response logic. I will try to include it inside this PR as including it in _future 5.0.0_ and again reinventing the API might be frustrating, as all code that uses `adafruit_httpserver` would have to be modifid again.

michalpokusa commented:
I believe current state of this PR is satisfying. All the changes, both added initially and ones added after creating PR are listed in the first post above, with EDIT next to them if they were arre/modified later. I do not plan any more features here, already there are plenty of new ones. 

This PR is a good foundation for implementing e.g. middleware or default handlers in the future, if the need for them comes.

Testing and code reviews or any other feedback would be appreciated.

FoamyGuy commented on examples/httpserver_simpletest_auto.py:
```diff
@@ -0,0 +1,23 @@
+
+
+@server.route("/")
+def base(request: Request):
```
This handler doesn't get called when you navigate to the device in a browser. For my device I'm loading `http://192.168.1.121/` but am getting a blank page without the message on it. In the dev tools on the browser I can see it's a 404 response. 

I was able to get it working by adding `[GET]` in the `@server.route` annotation arguments:

```
@server.route("/", [GET])
```

My assumption from these results is that methods argument for server route is functionally required. It doesn't raise an exception if you omit it, but will not ever get called since it won't match any methods. I did only basic testing on GET request so far though, perhaps the reality is more nuanced with other request types or scenarios. 

I think I am in favor of one of the following (but I don't have a strong opinion on which):

- Make `[GET]` the default value of the `methods` argument so that if the user doesn't declare it, they'll have default behavior that responds to GET requests. 

- Make the argument "fully required" and raise an exception if they do not pass this argument. This way the user won't have to troubleshoot the blank 404 responses if they end up in this situation. In this case, we'd also want to update any examples that have omitted this argument so that they'll run and serve responses as expected.

michalpokusa replied:
It currently defaults to `GET`, but there is in fact a error in parsing logic, I identified the problem, will fix soon.

Thanks for spotting that.

FoamyGuy replied:
Ahh, I see. I tested it successfully in the latest commits. Thank you.

FoamyGuy commented on adafruit_httpserver/server.py:
```diff
@@ -2,97 +2,220 @@
+        self._routes = _Routes()
         self._socket_source = socket_source
         self._sock = None
         self.root_path = root_path
```
While the current behavior around root_path is much improved from prior versions, I do think it's worth adding one additional warning to the user in the form of a print statement if they do set `root_path` to `/` 

Maybe something like this:

```
if self.root_path == "/":
    print("WARNING: Setting root_path to '/' will expose all files on your device through the webserver, including potentially sensitive files like settings.toml or secrets.py. Consider making a sub-directory on your device and using that for your root_path instead.")
```

michalpokusa replied:
I added the warning when `root_path` is set to `"/"` or `""`, which results in the same behavior as `"/"`.

FoamyGuy replied:
Nice! I tested the new warning successfully. Thanks for adding that. (And again for all of your work on this library!)

FoamyGuy commented on adafruit_httpserver/server.py:
```diff
@@ -2,97 +2,220 @@
+    def route(
+        self,
+        path: str,
+        methods: Union[str, Set[str]] = GET,
```
After tinkering a bit more I think this is one cog in the issue about required / default  methods argument value. The way this is typed, and the default value seem to indicate that it accepts either a single method string like `GET` or a set of them like `[GET, POST]`.

But it seems the single method string version does not actually work. 

If you use a route like this
```
@server.route("/hello", GET)
```
It does not get called when user navigates to `http://<server>/hello`.  If you change it to use `[GET]` instead for the argument then it does get called. 

I also verified that changing the default value here to `[GET]` instead of it's current value `GET` does allow you to start using routes with default arguments like:

```
@server.route("/hello")
```
And have it will be called and return it's response.


It seems like at least at one point the intention was to support either single or multiple. I think I'd lean toward just forcing multiple which is how it behaves now if I'm understanding correctly. If user wants to use the single one they can include their single item in the list. By not supporting both types it should make the code implementing the behavior a little easier because it can always just iterate over them and act upon them without needing to worry about which type was sent.

Requiring the list seems to be how Flask works as well if I am understanding the docs at the very bottom of this section: https://flask.palletsprojects.com/en/2.3.x/api/#url-route-registrations inside of the table for the details of the `**options` value

michalpokusa replied:
As above, this was purely the implementation issue, the intention was to support both single and multiple methods, so typing is ok, now it should be working fine.

Thank you for finding that, it was a very simple mistake that might confuse others, especially considering that there was no exception raised.

FoamyGuy commented on examples/httpserver_redirects.py:
```diff
@@ -0,0 +1,41 @@
+    """
+
+    if slug is None or not slug in REDIRECTS:
+        return Response(request, text="Unknown redirect", status=NOT_FOUND_404)
```
This Response is raising an exception when it attempts to init and therefore doesn't make it back to the browser correctly. 

The exception raised is:
```
Traceback (most recent call last):
  File "/lib/adafruit_httpserver/server.py", line 327, in poll
  File "/lib/adafruit_httpserver/server.py", line 266, in _handle_request
  File "/lib/adafruit_httpserver/route.py", line 154, in wrapped_handler
  File "code.py", line 37, in redirect_other
TypeError: unexpected keyword argument 'text'
```

michalpokusa replied:
Fixed, I must have missed it during refactoring. Nice catch.

michalpokusa commented:
Thank you for testing and merging! 👍
